docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+    }
+}
+```
+
+## Dependency Provider
+
+```
+[Organization]
+├── Zed
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Yves
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Glue
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Client
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+└── Service
+    └── [Module]        
+        ├── [Module]Config.php
+        └── [Module]DependencyProvider.php
+```
+
+**Description**
+
+Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 
+
+Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
+- The `provide` method is the highest level that holds only `add` calls without any additional logic.
+- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
+- The `get` method is the lowest level that sources the dependency.
+
+**Conventions**
+- All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).
+- Each dependency MUST be defined using a class constant.
+- Setting dependency using the `container::set()` MUST be paired with late-binding closure definition to decouple instantiation.
+- Dependencies MUST be wired through the target layer corresponding inherited method:
+```php
+public function provideCommunicationLayerDependencies(Container $container)
+public function provideBusinessLayerDependencies(Container $container)
+public function providePersistenceLayerDependencies(Container $container)
+public function provideDependencies(Container $container)
+public function provideBackendDependencies(Container $container)
+public function provideServiceLayerDependencies(Container $container)
+```
+- Only three type of methods CAN be defined, either `provide`, `get`, or `add`.
+```php
+public function provide*Dependencies(Container $container)
+```
+- `Provide` methods MUST call their parent `provide` method.
+- `Provide` methods MUST only call `add` methods.
+
+```php
+protected function add*(Container $container)
+```
+- `Add` methods MUST only introduce one dependency to the `Container` (a plugin-stack is considered one dependency in this respect).
+
+```php
+protected function get[Dependency](Container $container)
+protected function get[PluginInterfaceName]Plugins(Container $container)
+```
+
+<details><summary markdown='span'>Additional Conventions for Project Development</summary>
+- Constants can have any access modifier, they are not limited to `public`. <br/>
+- Methods can have any access modifier. <br/>
+- `Provide` methods can call other methods than `add`. <br/>
```

Comment:
Why does project development differ from the core development in this particular case?

```suggestion
- `Provide` methods can call other methods than `add`. <br/>
```

Reply:
it's the other way. can you tell me a reason why wouldn't it be allowed for project development?

Reply:
You are right, there is no reason that I could think of to have such restriction.

I am just a little bit confused for whatever reason we have this restriction for the core development, it may also benefit the project development.

Reply:
@bnymn single responsibility. it's simpler to read the code. but this is not something that we can force on project without reason.